### PR TITLE
PI6 Hackfest: Add support for local, Docker-based SQS/SNS to CMR

### PR DESCRIPTION
**Pull request by oubiwann**
This adds support to CMR for optionally running a local SQS/SNS clone (requires Docker). New features/capabilities include:
 * `lein` commands to start. stop, and restart an SQS/SNS container
 * configuration options that support the use of local SQS/SNS
 * updates to the messaging client to support custom endpoints
 * Updated docs (README) with instructions on using

Additionally:
 * utility functions for examining the system environment variables
 * refactoring dev system to match other services (with dedicated config ns)
 * commented out two assertions in a test that was intermittently failing and opened a ticket for these

### CODE REVIEW

File: message-queue-lib/src/cmr/message_queue/test/queue_broker_wrapper.clj [hunk 1]
```diff
@@ -113,13 +113,13 @@
   (let [{:keys [id-sequence-atom timeout?-atom queue-broker]} broker
         msg-id (swap! id-sequence-atom inc)
         tagged-msg (assoc msg :id msg-id)]
-
     ;; Set the initial state of the message to :initial
     (update-message-queue-history broker queue-name :enqueue tagged-msg :initial)
-
+    (debug "Updated message queue history")
```
Comment (chris-durbin) [inline message-queue-lib/src/cmr/message_queue/test/queue_broker_wrapper.clj]: That's a lot of debug... Time for `trace`. 😄

File: message-queue-lib/test/cmr/message_queue/test/queue/sqs.clj [hunk 2]
```diff
@@ -54,26 +74,74 @@
...
+
+(defn- recv-fn
+  "Returns a function to call on receive messages."
+  [receive-fns idx]
+  (if (< idx (count receive-fns))
+    (nth receive-fns idx)
+    #(throw (ExitException.))))
+
+(defn- fake-client-proxy
+  "Creates a proxy for AWSSQSClient. This prox does the following:
```
Comment (chris-durbin) [inline message-queue-lib/test/cmr/message_queue/test/queue/sqs.clj]: prox typo

File: message-queue-lib/test/cmr/message_queue/test/queue/sqs.clj [hunk 3]
```diff
@@ -54,26 +74,74 @@
...
+
+  * Returns a `GetQueueUrlRequest` proxy for calls to getQueueUrl.
+  * For calls to `receiveMessages`, it calls the functions in receive-fns
+    starting with the first function and then using the next function in the
+    list for each subsequent call until the list of receive functions is
+    exhausted. At that point subsequent calls to receiveMessages will throw an
+    ExitException to force the calling thread to exit. This is done to prevent
+    test worker threads from persisting after tests complete."
```
Comment (chris-durbin) [inline message-queue-lib/test/cmr/message_queue/test/queue/sqs.clj]: Thank you. I was actually looking to do exactly this the last time I ran the AWS tests, but ran out of time to figure out how to do it.

File: system-int-test/test/cmr/system_int_test/search/collection_humanized_search_test.clj [hunk 4]
```diff
@@ -108,10 +108,13 @@
   (search/refresh-collection-metadata-cache)
   (testing "Humanizer report batches"
     (let [report-lines (str/split (search/get-humanizers-report) #"\n")]
-      (is (= (count report-lines) (+ 2 (hrs/humanizer-report-collection-batch-size))))
-      (for [actual-line (rest report-lines)
-            n (inc hrs/humanizer-report-collection-batch-size)]
-        (is (= actual-line) (str "PROV1,C1200000001-PROV1,B,"n",AM-1,Terra"))))))
+      ;; XXX the following have been commented out due to unpredictable errors
```
Comment (chris-durbin) [inline system-int-test/test/cmr/system_int_test/search/collection_humanized_search_test.clj]: Yes, let's keep filing tickets for any intermittent test failures (and fix those tickets).

File: dev-system/src/cmr/dev_system/system.clj [hunk 5]
```diff
@@ -84,26 +82,29 @@
...
-  "Modifies the app server instances to configure web server options. Takes the system
-  and returns it with the updates made. Should be run a system before it is started"
+  "Modifies an app server instance to configure web server options, returning a
+  key/value pair that is the new (updated) app map."
+  [app-map]
+  (->> app-map
+       (map (fn [[app-name app-system]]
+              [app-name (update-app-web-server-options app-system)]))
+       (into {})))
+  ; (into {} (for [[app-name app-system] app-map]
```
Comment (oubiwann) [inline dev-system/src/cmr/dev_system/system.clj]: Whoops, these commented lines need to be removed.

File: message-queue-lib/src/cmr/message_queue/queue/sqs.clj [hunk 6]
```diff
@@ -244,16 +294,75 @@
 
 (defn- -get-queue-url
   "Returns the queue url for the given queue name."
-  [sqs-client queue-name]
+  [sqs-client normalized-lookup queue-name]
```
Comment (chris-durbin) [inline message-queue-lib/src/cmr/message_queue/queue/sqs.clj]: Looks like `normalized-lookup` was not needed after your debugging was finished. Can you remove it?

File: message-queue-lib/src/cmr/message_queue/queue/sqs.clj [hunk 7]
```diff
@@ -315,7 +426,9 @@
     [this queue-name msg]
     (let [msg (json/generate-string msg)
           queue-name (normalize-queue-name queue-name)
-          queue-url (get-queue-url sqs-client queue-name)]
+          queue-url (get-queue-url
```
Comment (chris-durbin) [inline message-queue-lib/src/cmr/message_queue/queue/sqs.clj]: Looks like we should back out this change.

File: dev-system/README.md [hunk 8]
```diff
@@ -28,6 +28,34 @@ Furthermore, there is a second (and optional) test runner you can use for
...
+* Run `lein start-sqs-sns`
+* From the shell where you will start the REPL, you will need the
+  `CMR_SNS_ENDPOINT` and `CMR_SQS_ENDPOINT` environment variables set;
+  in most cases you will want both of these set to `http://localhost:4100`
+* You will also need to set the env var
+  `CMR_SQS_EXTEND_POLICY_REMAINING_EXCHANGES` to `false` and to set the
+  standard AWS credential environment variables, `AWS_ACCESS_KEY_ID` and
+  `AWS_SECRET_ACCESS_KEY` (it doesn't matter what the actual values are).
+* For CMR to use the local SQS/SNS, it needs to have the
+  `CMR_DEV_SYSTEM_QUEUE_TYPE` environment variable set to "aws".
```
[the remark this answers is not in the document]
Comment (oubiwann) [reply dev-system/README.md]: That's actually what I was using at first, but I wasn't sure how this was used in CI, and I didn't want local SQS/SNS integration to be activated unintentionally ... so I opted for the much more explicit `CMR_DEV_SYSTEM_QUEUE_TYPE`.

I just took another look at dev-system's `user.clj` file and it looks like there were some logic gaps. I've pushed up some changes that should have the cases covered for using `CMR_DEV_SYSTEM_QUEUE_TYPE`.

File: message-queue-lib/src/cmr/message_queue/queue/sqs.clj (hunk 6 above)
Comment (oubiwann) [reply message-queue-lib/src/cmr/message_queue/queue/sqs.clj]: Will do.

File: message-queue-lib/src/cmr/message_queue/queue/sqs.clj [hunk 9]
```diff
@@ -59,11 +62,36 @@
...
+  When running SQS/SNS locally, however, to keep policy handling simple, we
+  do not want to extend the policies."
+  {:default true
+   :type Boolean})
+
+(defconfig sns-endpoint
+  "By default the SNS client will not explicitly set the SNS endpoint. This
+  configuration is provided for use in situations where explicitly setting
+  the SNS endpoint is desirable or required."
+  {:default nil
```
[the remark this answers is not in the document]
Comment (oubiwann) [reply message-queue-lib/src/cmr/message_queue/queue/sqs.clj]: Done.

File: message-queue-lib/src/cmr/message_queue/queue/sqs.clj (hunk 6 above)
Comment (oubiwann) [reply message-queue-lib/src/cmr/message_queue/queue/sqs.clj]: Done.

File: message-queue-lib/src/cmr/message_queue/queue/sqs.clj (hunk 7 above)
Comment (oubiwann) [reply message-queue-lib/src/cmr/message_queue/queue/sqs.clj]: Done.

File: message-queue-lib/test/cmr/message_queue/test/queue/sqs.clj (hunk 2 above)
Comment (oubiwann) [reply message-queue-lib/test/cmr/message_queue/test/queue/sqs.clj]: Fixed.

File: message-queue-lib/test/cmr/message_queue/test/queue/sqs.clj (hunk 3 above)
Comment (oubiwann) [reply message-queue-lib/test/cmr/message_queue/test/queue/sqs.clj]: 🙇

File: message-queue-lib/src/cmr/message_queue/queue/sqs.clj [hunk 10]
```diff
@@ -244,16 +294,75 @@
...
+  With the understanding, of course, that the configuration functions in the
+  example all have been created.
+
+  Note that since AWS service clients will set service defaults on their own,
+  it is best to define the default configuration value as nil, so as to avoid
+  unexpected client behaviour."
+  [client-obj & args]
+  (debug "Configuring client" client-obj)
+  (doseq [[config-key config-val] (partition 2 args)]
+    (set-aws-client-attr! client-obj config-key [config-val]))
```
[the remark this answers is not in the document]
Comment (oubiwann) [reply message-queue-lib/src/cmr/message_queue/queue/sqs.clj]: Yeah, it's "good" ol' OOP; the setter is 100% side-effecty. The functions I added return the object to be good FP citizens.

**Comment by oubiwann** [comment]
Thanks guys!

**Pull request merged**